OpenRA.Game/Primitives/PriorityQueue.cs

```diff
@@ -22,98 +22,145 @@ public interface IPriorityQueue<T>
 		T Pop();
 	}
 
-	public class PriorityQueue<T> : IPriorityQueue<T>
+	/// <summary>
+	/// Represents a collection of items that have a priority.
+	/// On pop, the item with the lowest priority value is removed.
+	/// </summary>
+	public sealed class PriorityQueue<T, TComparer> : IPriorityQueue<T> where TComparer : struct, IComparer<T>
 	{
-		readonly List<T[]> items;
-		readonly IComparer<T> comparer;
-		int level, index;
-
-		public PriorityQueue()
-			: this(Comparer<T>.Default) { }
-
-		public PriorityQueue(IComparer<T> comparer)
+		/// <summary>
+		/// Compares two items to determine their priority.
+		/// PERF: Using a struct allows the calls to be devirtualized.
+		/// </summary>
+		readonly TComparer comparer;
+
+		/// <summary>
+		/// A min-heap storing the items. An array divided into sub arrays called levels.
+		/// At each level the size of a level array doubles.
+		/// Elements at deeper levels always have higher priority than elements nearer to the root.
+		/// </summary>
+		T[] items;
```

Comment:
Oh hmm, in the current priority queue description it says that it pops the item with the lowest priority (closest to root). This inverse logic doesn't make much sense

Reply:
Changed `min-heap` to `binary min-heap` and `lowest priority` to `lowest priority value`.

The lowest priority value is correct - which suits our use case of wanting to return the lowest cost items first. Contrast a max heap where the largest values are nearer to the root. See https://en.wikipedia.org/wiki/Binary_heap

I'm hoping the "lowest priority value" helps make that less confusing - but I appreciate that it is confusing since the item we care about most which you might call the "highest priority item" is the one with the lowest value/cost.

If you have any wording suggestions am happy to apply them.

Reply:
Apart from adding the link to the file, I think this is fine. If you do it might be better served as a comment rather than summary